File: THCS/2022_2023/PhuYen_2223/Bai2/BAI2.py
def is_harmonic(A, B, start):
    """
    Hàm kiểm tra xem đoạn con B có đồng điệu với dãy A hay không.
    A là dãy mẫu, B là dãy cần kiểm tra, start là vị trí bắt đầu của đoạn con trong B.
    """
    k = B[start] - A[0]
    N = len(A)
    
    for i in range(1, N):
        if B[start + i] != A[i] + k:
            return False
    return True

def count_harmonic_sequences(A, B):
    """
    Hàm đếm số dãy con đồng điệu với dãy A trong dãy B.
    """
    N = len(A)
    M = len(B)
    count = 0
    
    i = 0
    while i <= M - N:
        if is_harmonic(A, B, i):
            count += 1
            i += N  # Bỏ qua các đoạn con rời nhau
        else:
            i += 1
    return count

# Đọc dữ liệu từ file BAI2.INP
with open("BAI2.INP", "r") as inp_file:
    # Đọc dòng đầu tiên chứa N và dãy A
    A = list(map(int, inp_file.readline().split()))[1:]  # Bỏ qua N, chỉ lấy dãy A
    
    # Đọc dòng thứ hai chứa dãy B
    B = list(map(int, inp_file.readline().split()))[1:]  # Bỏ qua M, chỉ lấy dãy B

# Đếm số dãy con đồng điệu
result = count_harmonic_sequences(A, B)

# Ghi kết quả vào file BAI2.OUT
with open("BAI2.OUT", "w") as out_file:
    out_file.write(str(result) + "\n")

File: THCS/2022_2023/PhuYen_2223/Bai2/BAI2.cpp
#include <iostream>
#include <fstream>
#include <vector>

using namespace std;

// Hàm kiểm tra xem đoạn con B có đồng điệu với dãy A hay không
bool isHarmonic(const vector<int>& A, const vector<int>& B, int start) {
    int k = B[start] - A[0];
    int N = A.size();
    
    for (int i = 1; i < N; i++) {
        if (B[start + i] != A[i] + k) {
            return false;
        }
    }
    return true;
}

int main() {
    // Mở file BAI2.INP để đọc dữ liệu
    ifstream inp_file("BAI2.INP");
    ofstream out_file("BAI2.OUT");

    if (!inp_file || !out_file) {
        cerr << "Không thể mở file!" << endl;
        return 1;
    }

    int N, M;
    // Đọc dãy A
    inp_file >> N;
    vector<int> A(N);
    for (int i = 0; i < N; i++) {
        inp_file >> A[i];
    }

    // Đọc dãy B
    inp_file >> M;
    vector<int> B(M);
    for (int i = 0; i < M; i++) {
        inp_file >> B[i];
    }

    int count = 0;
    // Kiểm tra từng đoạn con của B có đồng điệu với A hay không
    for (int i = 0; i <= M - N; i++) {
        if (isHarmonic(A, B, i)) {
            count++;
            i += N - 1; // Bỏ qua các đoạn con rời nhau
        }
    }

    // Ghi kết quả vào file BAI2.OUT
    out_file << count << endl;

    // Đóng file
    inp_file.close();
    out_file.close();

    return 0;
}
